File: linked-lists/7.c
/**
 * 7.c
 * Linked Lists
 * https://github.com/ashenm/dsa
 *
 * Implements Queue-like operations
 *
 * Ashen Gunaratne
 * mail@ashenm.ml
 *
 */

#include <stdio.h>
#include <stdlib.h>

typedef struct node
{
    int value;
    struct node *next;
} node;

// conjoins a new node
node* queue (node **head, int value);

// removes the first insert
node* dequeue (node **head);

// computes the node count
int range (node *head);

int main (void)
{

    node *head = NULL;

    queue(&head, 25);
    queue(&head, 50);

    printf("%i\n", range(head));

    free(dequeue(&head));

    node *elect = head;

    while (elect != NULL)
    {
        printf("%i\n", elect -> value);
        elect = elect -> next;
    }

    while (head != NULL)
    {
        elect = head;
        head = head -> next;
        free(elect);
    }

}

node* queue (node **head, int value)
{

    node *elect = *head;
    node *inset = malloc(sizeof(node));

    if (inset == NULL)
    {
        return NULL;
    }

    inset -> value = value;
    inset -> next = NULL;

    if (elect == NULL)
    {
        return *head = inset;
    }

    while (elect -> next != NULL)
    {
        elect = elect -> next;
    }

    elect -> next = inset;

    return *head;

}

node* dequeue (node **head)
{

    node *elect = *head;

    if (elect == NULL)
    {
        return NULL;
    }

    *head = elect -> next;

    return elect;

}

int range (node *head)
{

    int n = 0;

    while (head != NULL)
    {
        head = head -> next;
        n += 1;
    }

    return n;

}

// vim: set expandtab shiftwidth=4:

File: linked-lists/7.js
/**
 * 7.js
 * Linked Lists
 * https://github.com/ashenm/dsa
 *
 * Implements Queue-like operations
 *
 * Ashen Gunaratne
 * mail@ashenm.ml
 *
 */

'use strict';

class Node {

  constructor (value, next=null) {
    this.value = value;
    this.next = next;
  }

}

class LinkedList {

  constructor (head=null) {
    this.head = head;
  }

  // conjoins a new node
  queue (value) {

    const inset = new Node(value);

    if (!this.head) {
      return this.head = inset, this;
    }

    let elect = this.head;

    while (elect.next) {
      elect = elect.next;
    }

    elect.next = inset;

    return this;

  }

  // removes the firstmost insert
  dequeue () {

    const elect = this.head;

    if (elect) {
      this.head = elect.next;
    }

    return elect;

  }

  // computes the node count
  range () {

    let n = 0;
    let elect = this.head;

    while (elect) {
      elect = elect.next;
      n += 1;
    }

    return n;

  }

}

const main = function () {

  const list = new LinkedList();

  list.queue(25);
  list.queue(50);

  console.log(list.range());

  list.dequeue();

  for (let elect = list.head; elect; elect = elect.next) {
    console.log(elect.value);
  }

};

if (require.main === module) {
  main();
}

// vim: set expandtab shiftwidth=2:
